### PR TITLE
Always mark memory reference types as pointers.

### LINKED ISSUE #8489: Allow mapping to struct with dynamically-sized type to implement inherited function
Overriding functions defined in interfaces with the automatically-generated getters from public state variables usually work great. This was reported in a previous issue (#3514) and fixed.

However, when the return type is a `struct` that has a dynamically-sized type, like the `string` in the example below, the generated getter doesn't override properly, causing a compiler error.

```solidity
// sample.sol
pragma solidity 0.6.4;
pragma experimental ABIEncoderV2;

interface IA {
    struct Baz {
        address a;
        address b;
        string  c;
    }

    function bazMap(address) external returns (Baz memory);
}

contract A is IA {
    mapping (address => Baz) public override bazMap;
}
```

```sh
$ solc sample.sol  # it's version 0.6.4
Error: Overriding public state variable return types differ.
  --> sample.sol:15:5:
   |
15 |     mapping (address => Baz) public override bazMap;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: Overridden function is here:
  --> sample.sol:11:5:
   |
11 |     function bazMap(address) external returns (Baz memory);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Based on [the docs](https://solidity.readthedocs.io/en/v0.6.4/contracts.html#getter-functions), I might expect the function to override to be like:
```
function bazMap(address) external returns (address,address)
```
but that doesn't work either. 

Related to: #6337, #3514, #7521, #4244

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/8489

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -2564,10 +2564,14 @@ FunctionType::FunctionType(VariableDeclaration const& _varDecl):
...
 						continue;
-				m_returnParameterTypes.push_back(TypeProvider::withLocationIfReference(
-					DataLocation::Memory,
-					member.type
-				));
+				if (auto referenceType = dynamic_cast<ReferenceType const*>(member.type))
+					m_returnParameterTypes.push_back(TypeProvider::withLocation(
+						referenceType,
+						DataLocation::Memory,
+						true
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Please someone confirm, that I'm right in that this should in fact always be pointers under any circumstances :-).

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: The distinction between pointer and non-pointer is only relevant for storage types.

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Or phrased differently: Memory references should always be pointers.

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Ok, if that's so, then we should probably rather change ``copyForLocation``, ``withLocation``, ``withLocationIfReference``, etc., to always make the types pointers, if in memory...

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: But yeah, thinking about it, I don't know when it should be different, so I'll look into changing those parts instead.

File: test/libsolidity/gasTests/abiv2.sol [hunk 2]
```diff
@@ -14,9 +14,9 @@ contract C {
 }
 // ----
 // creation:
-//   codeDepositCost: 1120000
-//   executionCost: 1160
-//   totalCost: 1121160
+//   codeDepositCost: 1094400
+//   executionCost: 1134
+//   totalCost: 1095534
```
Comment (ekpyron) [inline test/libsolidity/gasTests/abiv2.sol:19]: My guess here is that we used to generate pointer and non-pointer versions of the same functions and those become unified with this change...

File: test/libsolidity/gasTests/abiv2_optimised.sol [hunk 3]
```diff
@@ -17,9 +17,9 @@ contract C {
 // optimize-yul: true
 // ----
 // creation:
-//   codeDepositCost: 603000
-//   executionCost: 638
-//   totalCost: 603638
+//   codeDepositCost: 610600
+//   executionCost: 645
+//   totalCost: 611245
```
Comment (ekpyron) [inline test/libsolidity/gasTests/abiv2_optimised.sol]: This *increasing* is weird, though, that I don't quite understand yet...

File: libsolidity/ast/Types.cpp [hunk 4]
```diff
@@ -1527,8 +1529,6 @@ string ReferenceType::identifierLocationSuffix() const
 		id += "_calldata";
 		break;
 	}
-	if (isPointer())
-		id += "_ptr";
```
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: Is this breaking? I could also *always* append ``_ptr`` for memory types, but both variants change the JSON ASTs one way or another...

File: test/libsolidity/gasTests/abiv2_optimised.sol (hunk 3 above)
Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2_optimised.sol]: Even weirder: if I generally add ``_ptr`` to ``identifierLocationSuffix`` for memory references, then this seems to decrease again...

File: test/libsolidity/gasTests/abiv2.sol (hunk 2 above)
Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2.sol:19]: Or we did superfluous conversions between pointer and non-pointer types or something like that...

File: test/libsolidity/gasTests/abiv2_optimised.sol (hunk 3 above)
Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2_optimised.sol]: I'll try to understand this by looking at the generated code...

Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2_optimised.sol]: Ok, this is weird... https://gist.github.com/ekpyron/8639d242e9a331bcefa8d02ed185f133
Special attention to the validator functions at the very end of the gist and their uses (or lack thereof)...

Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2_optimised.sol]: Ah ok... I think it's just inlining that is different... probably due to something being sorted by identifier somewhere...

**Comment by ekpyron** [comment]
I'm not entirely sure about this. I could roll it back to just change it for the return type of getters for now, that should be "safer" than the full change...

File: test/libsolidity/gasTests/abiv2.sol (hunk 2 above)
Comment (ekpyron) [reply test/libsolidity/gasTests/abiv2.sol:19]: https://gist.github.com/ekpyron/41f43961838274ad5f5dfb99c4b387e4
Looks identical short of function names... not sure why this changes...

**Comment by ekpyron** [comment]
I added another commit that *always* adds the ``_ptr`` suffix to identifiers for comparison. That commit can either be squashed into the change or dropped.

**Comment by chriseth** [comment]
I would like to see more asserts here.

**Comment by ekpyron** [comment]
Where exactly?

**Comment by chriseth** [comment]
in the constructor, for example

**Comment by chriseth** [comment]
Grabbing this.

File: libsolidity/ast/Types.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: This changes calldata, too. I think I'll revert this and only change the constructor.

File: test/libsolidity/syntaxTests/inheritance/override/public_var_override_mapping_to_dynamic_struct.sol [hunk 5]
```diff
@@ -0,0 +1,12 @@
+pragma experimental ABIEncoderV2;
+
+abstract contract C {
+    struct S {
+        uint256 x;
```
Comment (leonardoalt) [inline test/libsolidity/syntaxTests/inheritance/override/public_var_override_mapping_to_dynamic_struct.sol]: Mix of spaces and tabs

File: test/libsolidity/syntaxTests/inheritance/override/public_var_override_struct_with_memory_element.sol [hunk 6]
```diff
@@ -0,0 +1,13 @@
+pragma experimental ABIEncoderV2;
+
+abstract contract C {
+    struct S {
+        uint256 x;
+		string y;
```
Comment (leonardoalt) [inline test/libsolidity/syntaxTests/inheritance/override/public_var_override_struct_with_memory_element.sol:6]: Mix of spaces and tabs

**Review by leonardoalt — COMMENTED** [review]
Looks good, just these two comments

**Pull request merged**